File: apps/api/src/reception/reception.service.ts
import { BadRequestException, ConflictException, Injectable, NotFoundException } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { DataSource, EntityManager, Repository } from "typeorm";
import { QualityParameter, ReadingSource, TransactionStatus } from "@cc-mc/shared-types";
import { MilkReceptionTransaction } from "./entities/milk-reception-transaction.entity";
import { TransactionOverride } from "./entities/transaction-override.entity";
import { ChillingCentre } from "../centres/entities/chilling-centre.entity";
import { Source } from "../sources/entities/source.entity";
import { Vehicle } from "../vehicles/entities/vehicle.entity";
import { CentreAccessService } from "../rbac/centre-access.service";
import { AuditService } from "../audit/audit.service";
import { QualityRulesService } from "../quality-rules/quality-rules.service";
import { QualityValidationService } from "./quality-validation.service";
import type { RequestUser } from "../rbac/rbac.types";
import { CreateReceptionDto } from "./dto/create-reception.dto";
import { OverrideReceptionDto } from "./dto/override-reception.dto";

const REQUIRED_PARAMETERS = [QualityParameter.FAT, QualityParameter.SNF, QualityParameter.TEMPERATURE];

/**
 * The fields create() actually compares to decide "same logical payload"
 * for an idempotent retry (Checkpoint 5). Deliberately every business
 * field the caller controls - centreId/sourceId/vehicleId identify WHAT
 * was received, quantityKg/fat/snf/temperature identify the reading
 * itself. operatorUserId is NOT compared: two retries of the same
 * physical capture, from the gateway's perspective, always carry the
 * gateway's own service-account user id anyway (see the gateway
 * cloud-auth design), so it adds no discriminating signal and comparing
 * it would only risk a false conflict if that ever changed.
 */
type ReceptionPayloadFields = Pick<
  CreateReceptionDto,
  "centreId" | "sourceId" | "vehicleId" | "quantityKg" | "fat" | "snf" | "temperature"
>;

/**
 * Additive response shape for POST /reception (Checkpoint 5): every
 * existing field of MilkReceptionTransaction, unchanged, plus `outcome` -
 * "created" for a genuinely new row, "duplicate" for an idempotent retry
 * that returned an already-existing row. Both are HTTP 201 (Nest's POST
 * default, unchanged) - `outcome` is how a caller that cares (the
 * gateway's HttpCloudClient) tells them apart; the existing web UI simply
 * never reads this field, exactly like it never reads any other field it
 * doesn't use. A genuine payload conflict is NOT part of this type - see
 * create()'s ConflictException path.
 */
export type ReceptionCreateResult = MilkReceptionTransaction & { outcome: "created" | "duplicate" };

@Injectable()
export class ReceptionService {
  constructor(
    @InjectRepository(MilkReceptionTransaction)
    private readonly transactionRepo: Repository<MilkReceptionTransaction>,
    @InjectRepository(ChillingCentre) private readonly centreRepo: Repository<ChillingCentre>,
    @InjectRepository(Source) private readonly sourceRepo: Repository<Source>,
    @InjectRepository(Vehicle) private readonly vehicleRepo: Repository<Vehicle>,
    private readonly dataSource: DataSource,
    private readonly centreAccess: CentreAccessService,
    private readonly audit: AuditService,
    private readonly qualityRules: QualityRulesService,
    private readonly qualityValidation: QualityValidationService,
  ) {}

  async list(user: RequestUser) {
    if (user.centreAccess.allCentres) {
      return this.transactionRepo.find({ order: { receivedAt: "DESC" }, take: 100 });
    }
    if (user.centreAccess.centreIds.length === 0) return [];
    return this.transactionRepo.find({
      where: user.centreAccess.centreIds.map((centreId) => ({ centreId })),
      order: { receivedAt: "DESC" },
      take: 100,
    });
  }

  async getById(user: RequestUser, id: number) {
    const transaction = await this.transactionRepo.findOneBy({ id });
    if (!transaction) throw new NotFoundException("Reception transaction not found");
    this.centreAccess.assertCanAccess(user.centreAccess, transaction.centreId);
    return transaction;
  }

  async create(user: RequestUser, dto: CreateReceptionDto) {
    this.centreAccess.assertCanAccess(user.centreAccess, dto.centreId);

    const [centre, source, vehicle] = await Promise.all([
      this.centreRepo.findOneBy({ id: dto.centreId }),
      this.sourceRepo.findOneBy({ id: dto.sourceId }),
      this.vehicleRepo.findOneBy({ id: dto.vehicleId }),
    ]);

    if (!centre) throw new BadRequestException("Centre not found");
    if (!source || source.centreId !== dto.centreId) {
      throw new BadRequestException("Source not found for this centre");
    }
    if (source.status !== "ACTIVE") {
      throw new BadRequestException("Source is inactive");
    }
    if (!vehicle || vehicle.centreId !== dto.centreId) {
      throw new BadRequestException("Vehicle not found for this centre");
    }
    if (vehicle.status !== "ACTIVE") {
      throw new BadRequestException("Vehicle is inactive");
    }

    const rules = await this.qualityRules.resolveRulesForCentre(dto.centreId, REQUIRED_PARAMETERS);
    const validation = this.qualityValidation.validate(
      { fat: dto.fat, snf: dto.snf, temperature: dto.temperature },
      rules,
    );

    return this.dataSource.transaction(async (manager) => {
      const repo = manager.getRepository(MilkReceptionTransaction);

      // Checkpoint 5: a caller that supplied localIdempotencyKey (the
      // Local Device Gateway, always; the web UI, never) gets the
      // idempotent-retry-safe path. Every other field of this method is
      // untouched from Checkpoint 2 - the two paths share the same
      // validation, quality-rule resolution, and transaction-number
      // assignment; they only differ in how the row gets inserted.
      if (dto.localIdempotencyKey) {
        return this.createIdempotent(manager, repo, user, dto, centre, validation);
      }

      // Transaction number is assigned centrally, after the row has an id,
      // so two concurrent creations can never collide (see
      // docs/assumptions.md #transaction-numbering for why this differs
      // from the BRD's illustrative MCC-2026-000123 format).
      const created = await repo.save(
        repo.create({
          centreId: dto.centreId,
          sourceId: dto.sourceId,
          vehicleId: dto.vehicleId,
          operatorUserId: user.id,
          quantityKg: String(dto.quantityKg),
          fat: String(dto.fat),
          snf: String(dto.snf),
          temperature: String(dto.temperature),
          status: validation.status,
          readingSource: ReadingSource.MANUAL,
          reason: validation.reason,
          transactionNumber: `PENDING-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
          localIdempotencyKey: null,
        }),
      );

      created.transactionNumber = `${centre.code}-${created.id}`;
      const final = await repo.save(created);

      await this.audit.record(
        {
          userId: user.id,
          centreId: dto.centreId,
          action: "RECEPTION_CREATE",
          resourceType: "MilkReceptionTransaction",
          resourceId: String(final.id),
          oldValue: null,
          newValue: final,
          reason: validation.reason,
        },
        manager,
      );

      return { ...final, outcome: "created" as const };
    });
  }

  /**
   * The idempotent-retry-safe insert path (Checkpoint 5), used only when
   * the caller supplied a localIdempotencyKey. Uses the ALREADY-EXISTING
   * nullable-unique constraint on milk_reception_transactions.localIdempotencyKey
   * (see the InitSchema migration - this checkpoint adds no schema change)
   * as the actual correctness boundary, via `INSERT ... ON CONFLICT
   * ("localIdempotencyKey") DO NOTHING` - not an application-level
   * pre-check-then-insert, which would leave a race window between the
   * check and the insert. Postgres itself resolves the race: if two
   * transactions race to insert the same key, the SECOND one's INSERT
   * statement blocks until the FIRST one commits or rolls back; once it
   * unblocks, its own ON CONFLICT clause correctly sees whether the key
   * now exists (and skips) or doesn't (and proceeds) - see
   * docs/gateway-architecture.md's cloud-sync section and
   * test/reception-idempotency.e2e-spec.ts's concurrent test for the
   * proof. This one query is also why the manual conflict-detection path
   * below never needs its own locking: by the time it runs, Postgres has
   * already guaranteed the row it reads is the single, final, committed
   * winner.
   */
  private async createIdempotent(
    manager: EntityManager,
    repo: Repository<MilkReceptionTransaction>,
    user: RequestUser,
    dto: CreateReceptionDto,
    centre: ChillingCentre,
    validation: { status: TransactionStatus; reason: string | null },
  ): Promise<ReceptionCreateResult> {
    const insertResult = await repo
      .createQueryBuilder()
      .insert()
      .into(MilkReceptionTransaction)
      .values({
        centreId: dto.centreId,
        sourceId: dto.sourceId,
        vehicleId: dto.vehicleId,
        operatorUserId: user.id,
        quantityKg: String(dto.quantityKg),
        fat: String(dto.fat),
        snf: String(dto.snf),
        temperature: String(dto.temperature),
        status: validation.status,
        readingSource: ReadingSource.MANUAL,
        reason: validation.reason,
        transactionNumber: `PENDING-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
        localIdempotencyKey: dto.localIdempotencyKey,
      })
      .onConflict(`("localIdempotencyKey") DO NOTHING`)
      .returning(["id"])
      .execute();

    const insertedId: number | undefined = (insertResult.raw as Array<{ id: number }> | undefined)?.[0]?.id;

    if (insertedId !== undefined) {
      // Our INSERT actually landed a new row - finish exactly like the
      // non-idempotent path: assign the real transactionNumber now that
      // an id exists, save, and audit. Fetched fresh (not built from the
      // insert values) so `final` reflects exactly what the database has,
      // matching the non-idempotent path's behavior.
      const created = await repo.findOneByOrFail({ id: insertedId });
      created.transactionNumber = `${centre.code}-${created.id}`;
      const final = await repo.save(created);

      await this.audit.record(
        {
          userId: user.id,
          centreId: dto.centreId,
          action: "RECEPTION_CREATE",
          resourceType: "MilkReceptionTransaction",
          resourceId: String(final.id),
          oldValue: null,
          newValue: final,
          reason: validation.reason,
        },
        manager,
      );

      return { ...final, outcome: "created" as const };
    }

    // Our INSERT was skipped - a row with this localIdempotencyKey already
    // exists and (per the comment above) is guaranteed committed and
    // visible to us right now. Compare payloads: same payload is a safe,
    // audit-free replay; a different payload is a caller bug surfaced
    // loudly (see ReceptionPayloadFields's doc comment) - either way,
    // nothing new is written, and NO second audit record is created.
    const existing = await repo.findOneByOrFail({ localIdempotencyKey: dto.localIdempotencyKey });
    const conflicts = this.findPayloadConflicts(existing, dto);

    if (conflicts.length > 0) {
      throw new ConflictException({
        message:
          `localIdempotencyKey "${dto.localIdempotencyKey}" was already used for a transaction with ` +
          `different data (conflicting field(s): ${conflicts.join(", ")}). Retrying with the same ` +
          `idempotency key must resubmit the SAME logical payload - this looks like a caller bug, not ` +
          `a legitimate retry.`,
        conflictingFields: conflicts,
        existingTransactionId: existing.id,
      });
    }

    return { ...existing, outcome: "duplicate" as const };
  }

  /**
   * Exact-value comparison between an already-persisted transaction and an
   * incoming request claiming the same localIdempotencyKey - mirrors the
   * gateway's own findPayloadConflicts() (apps/gateway/src/storage/
   * sqlite-local-storage.ts) so both halves of the same idempotency
   * contract agree on what "the same payload" means. Numeric columns are
   * compared via Number(...) rather than string equality because they are
   * persisted as fixed-precision decimal strings (e.g. "45.50") - a
   * caller resubmitting `45.5` must not be treated as a conflict merely
   * because of trailing-zero formatting.
   */
  private findPayloadConflicts(existing: MilkReceptionTransaction, incoming: ReceptionPayloadFields): string[] {
    const conflicts: string[] = [];
    if (existing.centreId !== incoming.centreId) conflicts.push("centreId");
    if (existing.sourceId !== incoming.sourceId) conflicts.push("sourceId");
    if (existing.vehicleId !== incoming.vehicleId) conflicts.push("vehicleId");
    if (Number(existing.quantityKg) !== incoming.quantityKg) conflicts.push("quantityKg");
    if (Number(existing.fat) !== incoming.fat) conflicts.push("fat");
    if (Number(existing.snf) !== incoming.snf) conflicts.push("snf");
    if (Number(existing.temperature) !== incoming.temperature) conflicts.push("temperature");
    return conflicts;
  }

  async override(user: RequestUser, id: number, dto: OverrideReceptionDto) {
    const transaction = await this.transactionRepo.findOneBy({ id });
    if (!transaction) throw new NotFoundException("Reception transaction not found");
    this.centreAccess.assertCanAccess(user.centreAccess, transaction.centreId);

    if (transaction.status !== TransactionStatus.HOLD) {
      throw new BadRequestException("Only a transaction currently on HOLD can be overridden");
    }

    return this.dataSource.transaction(async (manager) => {
      const txRepo = manager.getRepository(MilkReceptionTransaction);
      const overrideRepo = manager.getRepository(TransactionOverride);

      transaction.status = dto.newStatus;
      transaction.reason = dto.reason;
      const updated = await txRepo.save(transaction);

      await overrideRepo.save(
        overrideRepo.create({
          transactionId: id,
          originalStatus: TransactionStatus.HOLD,
          newStatus: dto.newStatus,
          performedByUserId: user.id,
          reason: dto.reason,
        }),
      );

      await this.audit.record(
        {
          userId: user.id,
          centreId: transaction.centreId,
          action: "RECEPTION_OVERRIDE",
          resourceType: "MilkReceptionTransaction",
          resourceId: String(id),
          oldValue: { status: TransactionStatus.HOLD },
          newValue: { status: dto.newStatus },
          reason: dto.reason,
        },
        manager,
      );

      return updated;
    });
  }
}

File: src/CCMC.Cloud.Application/Reception/ReceptionService.cs
using CCMC.Cloud.Application.Audit;
using CCMC.Cloud.Application.Auth;
using CCMC.Cloud.Application.Common;
using CCMC.Cloud.Domain.Entities;
using CCMC.Cloud.Domain.Enums;
using CCMC.Cloud.Domain.Services;
using CCMC.Cloud.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Npgsql;

namespace CCMC.Cloud.Application.Reception;

public sealed record CreateReceptionCommand(
    int CentreId, int SourceId, int VehicleId,
    decimal QuantityKg, decimal Fat, decimal Snf, decimal Temperature,
    string? LocalIdempotencyKey,
    TransactionStatus? Status = null,
    decimal? Clr = null, decimal? Water = null, decimal? Protein = null, string? RawAnalyserPayload = null,
    decimal? Rate = null, decimal? Amount = null);

public enum CreateReceptionOutcome { Created, Duplicate }

public sealed record CreateReceptionResult(CreateReceptionOutcome Outcome, MilkReceptionTransaction Transaction);

public sealed record OverrideReceptionCommand(TransactionStatus NewStatus, string Reason);

/// <summary>
/// The cloud's authoritative reception-mutation logic. CreateAsync mirrors
/// the exact idempotency contract the Windows client's HttpCloudApiClient/
/// SyncEngineService already assume (context.md/CLAUDE.md "API Contract" -
/// verified against CCMC.Contracts.Dtos.CreateReceptionRequestDto/
/// ReceptionTransactionDto and CCMC.Infrastructure.Sync.HttpResponseClassifier):
/// a request with a localIdempotencyKey that already exists returns the
/// existing row unchanged (Duplicate, still 201) if the payload matches, or
/// throws IdempotencyConflictException (409) naming every conflicting field
/// if it does not - never a silent overwrite, never a second create.
///
/// The actual correctness boundary is the database's own UNIQUE constraint
/// (ix_milk_reception_transactions_local_idempotency_key), not an
/// application-level "check then insert": the insert is attempted directly,
/// and a Postgres unique-violation on that specific index is what identifies
/// a genuine duplicate - Postgres itself serializes two concurrent inserts
/// racing for the same key, so there is no window for a query, then a
/// insert to see stale state.
/// </summary>
public sealed class ReceptionService(
    CcmcDbContext db,
    IAuditService audit,
    ILogger<ReceptionService> logger)
{
    private const string LocalIdempotencyKeyConstraintName = "ix_milk_reception_transactions_local_idempotency_key";

    private static readonly QualityParameter[] RequiredParameters =
        [QualityParameter.Fat, QualityParameter.Snf, QualityParameter.Temperature];

    public async Task<MilkReceptionTransaction> GetByIdAsync(RequestUser user, int id, CancellationToken cancellationToken)
    {
        var transaction = await db.MilkReceptionTransactions.AsNoTracking().SingleOrDefaultAsync(t => t.Id == id, cancellationToken)
            ?? throw new NotFoundException($"Reception transaction {id} not found.");
        CentreAccessGuard.AssertCanAccess(user.CentreAccess, transaction.CentreId);
        return transaction;
    }

    public async Task<IReadOnlyList<MilkReceptionTransaction>> ListAsync(RequestUser user, CancellationToken cancellationToken)
    {
        var query = db.MilkReceptionTransactions.AsNoTracking().OrderByDescending(t => t.ReceivedAt).Take(100);

        if (user.CentreAccess.AllCentres)
        {
            return await query.ToListAsync(cancellationToken);
        }

        if (user.CentreAccess.CentreIds.Count == 0) return [];

        return await query.Where(t => user.CentreAccess.CentreIds.Contains(t.CentreId)).ToListAsync(cancellationToken);
    }

    public async Task<CreateReceptionResult> CreateAsync(RequestUser user, CreateReceptionCommand cmd, CancellationToken cancellationToken)
    {
        CentreAccessGuard.AssertCanAccess(user.CentreAccess, cmd.CentreId);

        var centre = await db.ChillingCentres.AsNoTracking().SingleOrDefaultAsync(c => c.Id == cmd.CentreId, cancellationToken)
            ?? throw new ValidationException("Centre not found.");

        var source = await db.Sources.AsNoTracking().SingleOrDefaultAsync(s => s.Id == cmd.SourceId, cancellationToken);
        if (source is null || source.CentreId != cmd.CentreId) throw new ValidationException("Source not found for this centre.");
        if (source.Status != RecordStatus.Active) throw new ValidationException("Source is inactive.");

        var vehicle = await db.Vehicles.AsNoTracking().SingleOrDefaultAsync(v => v.Id == cmd.VehicleId, cancellationToken);
        if (vehicle is null || vehicle.CentreId != cmd.CentreId) throw new ValidationException("Vehicle not found for this centre.");
        if (vehicle.Status != RecordStatus.Active) throw new ValidationException("Vehicle is inactive.");

        if (cmd.Status == TransactionStatus.Rejected)
        {
            // Matches the client-side invariant exactly (TransactionStatus's own
            // doc comment, both sides): Rejected is only ever reached via an
            // override of a prior Hold, never a direct create.
            throw new ValidationException("A reception cannot be created directly with status Rejected - reject an existing HOLD via the override endpoint instead.");
        }

        var rules = await ResolveRulesAsync(cmd.CentreId, cancellationToken);
        var validation = QualityValidationService.Validate(new QualityReadingInput(cmd.Fat, cmd.Snf, cmd.Temperature), rules);

        // If the caller (the Windows client, going forward) already made an
        // explicit human ACCEPT/HOLD decision, that decision is authoritative -
        // it may legitimately differ from this automatic suggestion (e.g. the
        // operator judged the analyser's Water/Protein reading unacceptable
        // even though Fat/Snf/Temperature alone are in range, or overrode a
        // borderline HOLD). The automatic suggestion is still computed and
        // still recorded (dormant, not deleted - see CLAUDE.md/STATUS.md "Milk
        // Analyser + Quality Decision Flow"). A caller that omits Status
        // (none exists today, but the field is optional for backward
        // compatibility) gets the original, fully-automatic behavior unchanged.
        var decidedStatus = cmd.Status ?? validation.Status;
        var reason = cmd.Status is { } decided && decided != validation.Status
            ? (validation.Reason is null
                ? $"Operator decision ({decided}) overrides the automatic quality suggestion of {validation.Status}."
                : $"Operator decision ({decided}) overrides the automatic quality suggestion of {validation.Status}: {validation.Reason}")
            : validation.Reason;

        var now = DateTimeOffset.UtcNow;
        var entity = new MilkReceptionTransaction
        {
            // Placeholder, replaced with {centreCode}-{id} once the row has an id (below) -
            // guarantees uniqueness by construction, no cross-request numbering race.
            TransactionNumber = $"PENDING-{Guid.NewGuid():N}",
            CentreId = cmd.CentreId,
            SourceId = cmd.SourceId,
            VehicleId = cmd.VehicleId,
            OperatorUserId = user.Id,
            QuantityKg = cmd.QuantityKg,
            Fat = cmd.Fat,
            Snf = cmd.Snf,
            Temperature = cmd.Temperature,
            Clr = cmd.Clr,
            Water = cmd.Water,
            Protein = cmd.Protein,
            RawAnalyserPayload = cmd.RawAnalyserPayload,
            // Trusted verbatim from the client, same treatment as Fat/Snf/Clr/
            // Water/Protein - see MilkReceptionTransaction.Rate's doc comment
            // for why the cloud never recomputes these independently.
            Rate = cmd.Rate,
            Amount = cmd.Amount,
            Status = decidedStatus,
            ReadingSource = ReadingSource.Manual,
            Reason = reason,
            LocalIdempotencyKey = cmd.LocalIdempotencyKey,
            ReceivedAt = now,
            CreatedAt = now,
            UpdatedAt = now,
        };

        if (string.IsNullOrEmpty(cmd.LocalIdempotencyKey))
        {
            return await InsertAndFinalizeAsync(user, centre, entity, reason, cancellationToken);
        }

        await using var transaction = await db.Database.BeginTransactionAsync(cancellationToken);
        try
        {
            var result = await InsertAndFinalizeAsync(user, centre, entity, reason, cancellationToken);
            await transaction.CommitAsync(cancellationToken);
            return result;
        }
        catch (DbUpdateException ex) when (IsLocalIdempotencyKeyViolation(ex))
        {
            await transaction.RollbackAsync(cancellationToken);
            db.ChangeTracker.Clear();

            // The insert was rejected by the database's own unique constraint - a row
            // with this key is guaranteed already committed and visible to us now.
            var existing = await db.MilkReceptionTransactions.AsNoTracking()
                .SingleAsync(t => t.LocalIdempotencyKey == cmd.LocalIdempotencyKey, cancellationToken);

            var conflicts = FindConflicts(existing, cmd);
            if (conflicts.Count > 0)
            {
                throw new IdempotencyConflictException(
                    $"localIdempotencyKey \"{cmd.LocalIdempotencyKey}\" was already used for a transaction with different data " +
                    $"(conflicting field(s): {string.Join(", ", conflicts)}). Retrying with the same idempotency key must resubmit " +
                    "the SAME logical payload - this looks like a caller bug, not a legitimate retry.",
                    conflicts, existing.Id);
            }

            logger.LogInformation("Idempotent retry for localIdempotencyKey {Key} - returning existing transaction {Id}.", cmd.LocalIdempotencyKey, existing.Id);
            return new CreateReceptionResult(CreateReceptionOutcome.Duplicate, existing);
        }
    }

    private async Task<CreateReceptionResult> InsertAndFinalizeAsync(
        RequestUser user, ChillingCentre centre, MilkReceptionTransaction entity, string? reason,
        CancellationToken cancellationToken)
    {
        db.MilkReceptionTransactions.Add(entity);
        await db.SaveChangesAsync(cancellationToken); // assigns entity.Id, or throws DbUpdateException on unique violation

        entity.TransactionNumber = $"{centre.Code}-{entity.Id}";

        await audit.RecordAsync(
            new AuditEntry(user.Id, entity.CentreId, "RECEPTION_CREATE", "MilkReceptionTransaction", entity.Id.ToString(), Reason: reason),
            cancellationToken);

        await db.SaveChangesAsync(cancellationToken);

        logger.LogInformation(
            "Reception created: id={Id} transactionNumber={TransactionNumber} status={Status} centre={CentreId}",
            entity.Id, entity.TransactionNumber, entity.Status, entity.CentreId);

        return new CreateReceptionResult(CreateReceptionOutcome.Created, entity);
    }

    public async Task<MilkReceptionTransaction> OverrideAsync(RequestUser user, int transactionId, OverrideReceptionCommand cmd, CancellationToken cancellationToken)
    {
        if (cmd.NewStatus is not (TransactionStatus.Accepted or TransactionStatus.Rejected))
        {
            throw new ValidationException("newStatus must be Accepted or Rejected.");
        }

        await using var tx = await db.Database.BeginTransactionAsync(cancellationToken);

        var transaction = await db.MilkReceptionTransactions.SingleOrDefaultAsync(t => t.Id == transactionId, cancellationToken)
            ?? throw new NotFoundException($"Reception transaction {transactionId} not found.");

        CentreAccessGuard.AssertCanAccess(user.CentreAccess, transaction.CentreId);

        if (transaction.Status != TransactionStatus.Hold)
        {
            // No idempotency key exists for this endpoint in the current client
            // contract (documented gap - see CLAUDE.md "Architecture Decisions" /
            // "Known Gaps"): a retried override request cannot be distinguished
            // from a genuine second attempt, so this is always a terminal
            // conflict, never silently treated as "already succeeded."
            throw new BusinessConflictException("Only a transaction currently on HOLD can be overridden.");
        }

        var originalStatus = transaction.Status;
        transaction.Status = cmd.NewStatus;
        transaction.Reason = cmd.Reason;
        transaction.UpdatedAt = DateTimeOffset.UtcNow;

        db.TransactionOverrides.Add(new TransactionOverride
        {
            TransactionId = transactionId,
            OriginalStatus = originalStatus,
            NewStatus = cmd.NewStatus,
            PerformedByUserId = user.Id,
            Reason = cmd.Reason,
            CreatedAt = DateTimeOffset.UtcNow,
        });

        await audit.RecordAsync(
            new AuditEntry(
                user.Id, transaction.CentreId, "RECEPTION_OVERRIDE", "MilkReceptionTransaction", transactionId.ToString(),
                OldValueJson: $"{{\"status\":\"{originalStatus}\"}}", NewValueJson: $"{{\"status\":\"{cmd.NewStatus}\"}}", Reason: cmd.Reason),
            cancellationToken);

        await db.SaveChangesAsync(cancellationToken);
        await tx.CommitAsync(cancellationToken);

        logger.LogInformation(
            "Reception overridden: id={Id} {OriginalStatus} -> {NewStatus} by user {UserId}",
            transactionId, originalStatus, cmd.NewStatus, user.Id);

        return transaction;
    }

    private async Task<IReadOnlyDictionary<QualityParameter, ResolvedQualityRule>> ResolveRulesAsync(int centreId, CancellationToken cancellationToken)
    {
        var candidates = await db.QualityRules.AsNoTracking()
            .Where(r => RequiredParameters.Contains(r.Parameter) && (r.CentreId == centreId || r.CentreId == null))
            .ToListAsync(cancellationToken);

        var result = new Dictionary<QualityParameter, ResolvedQualityRule>();
        foreach (var parameter in RequiredParameters)
        {
            // Centre-specific overrides global.
            var rule = candidates.FirstOrDefault(r => r.Parameter == parameter && r.CentreId == centreId)
                       ?? candidates.FirstOrDefault(r => r.Parameter == parameter && r.CentreId is null);
            if (rule is not null)
            {
                result[parameter] = new ResolvedQualityRule(rule.MinValue, rule.MaxValue);
            }
        }
        return result;
    }

    /// <summary>
    /// Mirrors the exact comparison the Windows client's own local idempotency
    /// check already uses (context.md/CLAUDE.md - both sides must agree on
    /// what "the same payload" means): centreId/sourceId/vehicleId/quantityKg/
    /// fat/snf/temperature. OperatorUserId is deliberately NOT compared - see
    /// CreateReceptionCommand's absence of that field: the cloud always
    /// attributes a synced reception to whichever authenticated user's token
    /// performed the sync, which may legitimately differ between the original
    /// attempt and a retry (e.g. after a re-login), without that being a real conflict.
    /// </summary>
    private static List<string> FindConflicts(MilkReceptionTransaction existing, CreateReceptionCommand incoming)
    {
        var conflicts = new List<string>();
        if (existing.CentreId != incoming.CentreId) conflicts.Add(nameof(existing.CentreId));
        if (existing.SourceId != incoming.SourceId) conflicts.Add(nameof(existing.SourceId));
        if (existing.VehicleId != incoming.VehicleId) conflicts.Add(nameof(existing.VehicleId));
        if (existing.QuantityKg != incoming.QuantityKg) conflicts.Add(nameof(existing.QuantityKg));
        if (existing.Fat != incoming.Fat) conflicts.Add(nameof(existing.Fat));
        if (existing.Snf != incoming.Snf) conflicts.Add(nameof(existing.Snf));
        if (existing.Temperature != incoming.Temperature) conflicts.Add(nameof(existing.Temperature));
        if (incoming.Status is { } status && existing.Status != status) conflicts.Add(nameof(existing.Status));
        return conflicts;
    }

    private static bool IsLocalIdempotencyKeyViolation(DbUpdateException ex) =>
        ex.InnerException is PostgresException { SqlState: PostgresErrorCodes.UniqueViolation } pg &&
        pg.ConstraintName == LocalIdempotencyKeyConstraintName;
}
